src/renderer/components/input/input_validators.ts

```diff
@@ -4,37 +4,59 @@
  */
 
 import type { InputProps } from "./input";
-import type { ReactNode } from "react";
+import type React from "react";
 import fse from "fs-extra";
 import { TypedRegEx } from "typed-regex";
 
-export class AsyncInputValidationError extends Error {}
+export type InputValidationResult<IsAsync extends boolean> =
+  IsAsync extends true
+    ? Promise<void>
+    : boolean;
 
-export type InputValidator<IsAsync extends boolean> = {
+export type InputValidation<IsAsync extends boolean, RequireProps extends boolean> = (
+  RequireProps extends true
+    ? (value: string, props: InputProps) => InputValidationResult<IsAsync>
+    : (value: string, props?: InputProps) => InputValidationResult<IsAsync>
+);
+
+export type SyncValidationMessage<RequireProps extends boolean> = React.ReactNode | (
+  RequireProps extends true
+    ? (value: string, props: InputProps) => React.ReactNode
+    : (value: string, props?: InputProps) => React.ReactNode
+);
+
+export type InputValidator<IsAsync extends boolean = boolean, RequireProps extends boolean = boolean> = {
   /**
    * Filters itself based on the input props
    */
   condition?: (props: InputProps) => any;
 } & (
   IsAsync extends false
     ? {
-      validate: (value: string, props: InputProps) => boolean;
-      message: ReactNode | ((value: string, props: InputProps) => ReactNode | string);
+      validate: InputValidation<false, RequireProps>;
+      message: SyncValidationMessage<RequireProps>;
       debounce?: undefined;
     }
     : {
       /**
-       * If asyncronous then the rejection message is the error message
-       *
-       * This function MUST reject with an instance of {@link AsyncInputValidationError}
+       * The validation message maybe either specified from the `message` field (higher priority)
+       * or if that is not provided then the message will retrived from the rejected with value
        */
-      validate: (value: string, props: InputProps) => Promise<void>;
-      message?: undefined;
+      validate: InputValidation<true, RequireProps>;
+      message?: SyncValidationMessage<RequireProps>;
       debounce: number;
     }
 );
 
-export function inputValidator<IsAsync extends boolean = false>(validator: InputValidator<IsAsync>): InputValidator<IsAsync> {
+export function asyncInputValidator(validator: InputValidator<true, false>): InputValidator<true, false> {
+  return validator;
+}
+
+export function inputValidator(validator: InputValidator<false, false>): InputValidator<false, false> {
+  return validator;
+}
```

Comment:
I don't get how these two different.. just the name? They seems to get the same input, and return the same output.

Reply:
ah, now I found the diff, `InputValidator<false, false>` vs `InputValidator<true, false>`